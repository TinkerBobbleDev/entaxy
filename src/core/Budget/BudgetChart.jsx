import React, { useState } from 'react'
import { useSelector } from 'react-redux'
import { format, startOfMonth } from 'date-fns'
import {
  LineChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
  Legend,
  ResponsiveContainer
} from 'recharts'
import { currencyFormatter } from '../../util/stringFormatter'

const BudgetChart = () => {
  const {
    settings,
    transactions,
    budget,
    formatCurrency
  } = useSelector(state => ({
    settings: state.settings,
    transactions: state.transactions,
    budget: state.budget,
    formatCurrency: currencyFormatter(state.settings.locale, state.settings.currency)

  }))

  const usedCategories = [...new Set(Object.values(budget.rules).map(rules => (
    budget.categoriesById[rules.categoryId]
  )))]
  const initialState = usedCategories.reduce(
    (res, cat) => ({ ...res, [cat]: 1 }),
    {}
  )
  const [opacity, setOpacity] = useState(initialState)

  const byMonth = {}
  transactions.list.forEach((transaction) => {
    const dateKey = startOfMonth(transaction.createdAt).getTime()
    if (transaction.categoryId !== undefined) {
      const categoryName = budget.categoriesById[transaction.categoryId].name
      if (byMonth[dateKey] === undefined) {
        byMonth[dateKey] = usedCategories.reduce((res, cat) => ({ ...res, [cat.name]: 0 }), {})
      }
      if (byMonth[dateKey][categoryName] === undefined) {
        byMonth[dateKey][categoryName] = 0
      }
      byMonth[dateKey][categoryName] += Math.abs(transaction.amount)
    }
  })
  const data = Object.keys(byMonth).sort((a, b) => a - b).map(date => ({
    date: format(parseInt(date, 10), 'MMM YYYY'),
    ...byMonth[date]
  }))
  const margin = {
    top: 5,
    right: 20,
    bottom: 5,
    left: 20
  }

  const handleMouseEnter = (obj) => {
    setOpacity(
      usedCategories.reduce((res, category) => ({
        ...res,
        [category.name]: category.name === obj.dataKey ? 1 : 0.3
      }), {})
    )
  }
  const handleMouseLeave = () => {
    setOpacity(
      usedCategories.reduce((res, category) => ({
        ...res,
        [category.name]: 1
      }), {})
    )
  }

  return (
    <ResponsiveContainer>
      <LineChart data={data} margin={margin}>
        { usedCategories.map(category => (
          <Line
            type="monotone"
            dataKey={category.name}
            key={category.name}
            stroke={category.colour}
            strokeOpacity={opacity[category.name]}
            strokeWidth={2}
          />
        ))}
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="date" />
        <YAxis label={{ value: settings.currency, angle: -90, position: 'insideLeft' }} />
        <Tooltip
          formatter={(value, name) => [formatCurrency(value), name]}
        />
        <Legend
          layout="vertical"
          verticalAlign="top"
          align="right"
          wrapperStyle={{ paddingLeft: '40px' }}
          onMouseEnter={handleMouseEnter}
          onMouseLeave={handleMouseLeave}
        />
      </LineChart>
    </ResponsiveContainer>
  )
}

export default BudgetChart
