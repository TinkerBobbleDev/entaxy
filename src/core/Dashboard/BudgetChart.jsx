import React, { useState } from 'react'
import { useSelector } from 'react-redux'
import { format, startOfMonth } from 'date-fns'
import {
  LineChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
  Legend,
  ResponsiveContainer
} from 'recharts'
import { currencyFormatter } from '../../util/stringFormatter'

const BudgetChart = () => {
  const {
    settings,
    transactions,
    categories,
    colours,
    formatCurrency
  } = useSelector(state => ({
    settings: state.settings,
    transactions: state.transactions,
    categories: [...new Set(Object.values(state.budget.rules).map(rules => rules.category))],
    colours: state.budget.colours,
    formatCurrency: currencyFormatter(state.settings.locale, state.settings.currency)

  }))

  const initialState = categories.reduce(
    (res, cat) => ({ ...res, [cat]: 1 }),
    {}
  )
  const [opacity, setOpacity] = useState(initialState)

  const byMonth = {}
  transactions.list.forEach((transaction) => {
    const dateKey = startOfMonth(transaction.createdAt)
    if (transaction.category !== undefined) {
      if (byMonth[dateKey] === undefined) {
        byMonth[dateKey] = categories.reduce((res, cat) => ({ ...res, [cat]: 0 }), {})
      }
      if (byMonth[dateKey][transaction.category] === undefined) {
        byMonth[dateKey][transaction.category] = 0
      }
      byMonth[dateKey][transaction.category] += Math.abs(transaction.amount)
    }
  })
  const data = Object.keys(byMonth).sort((a, b) => a - b).map(date => ({
    date: format(date, 'MMM YYYY'),
    ...byMonth[date]
  }))
  const margin = {
    top: 5,
    right: 20,
    bottom: 5,
    left: 20
  }

  const handleMouseEnter = (obj) => {
    setOpacity(
      categories.reduce((res, value) => ({
        ...res,
        [value]: value === obj.dataKey ? 1 : 0.3
      }), {})
    )
  }
  const handleMouseLeave = () => {
    setOpacity(
      categories.reduce((res, value) => ({
        ...res,
        [value]: 1
      }), {})
    )
  }

  return (
    <ResponsiveContainer>
      <LineChart data={data} margin={margin}>
        { categories.map(category => (
          <Line
            type="monotone"
            dataKey={category}
            key={category}
            stroke={colours[category]}
            strokeOpacity={opacity[category]}
            strokeWidth={2}
          />
        ))}
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="date" />
        <YAxis label={{ value: settings.currency, angle: -90, position: 'insideLeft' }} />
        <Tooltip
          formatter={(value, name) => [formatCurrency(value), name]}
        />
        <Legend
          layout="vertical"
          verticalAlign="top"
          align="right"
          wrapperStyle={{ paddingLeft: '40px' }}
          onMouseEnter={handleMouseEnter}
          onMouseLeave={handleMouseLeave}
        />
      </LineChart>
    </ResponsiveContainer>
  )
}

export default BudgetChart
