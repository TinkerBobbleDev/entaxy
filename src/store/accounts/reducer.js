/* eslint-disable no-case-declarations */
import types from './types'

export const initialState = {
  byId: {}, // {id1: {id: id1, name: 'account1', institution: 'TD'}, id2: {...}}
  byInstitution: {} // {TD: {accountIds: [id1, id2], balance: 100, apiKey: XYZ}, BMO: {...}}
}

export const groupByInstitution = ({ byId, byInstitution }) => {
  // Group sorted accounts by institution
  // results in something like:
  // {
  //   TD: {accountIds: [id1, id2], balance: 100},
  //   BMO: {accountIds: [id3], balance: 200}
  // }
  const groupedAccountIds = Object.values(byId)
    // Sort accounts by name
    .sort((a, b) => (a.name > b.name))
    // Group them by institution
    .reduce((result, account) => {
      const {
        id,
        institution,
        groupId,
        currentBalance
      } = account

      // Merge existing institution data
      const existingInstitutionData = byInstitution[institution] || { groups: {} }
      const currentInstitutionData = (result[institution] || { balance: 0, groups: {} })
      const newInstitutionData = { ...existingInstitutionData, ...currentInstitutionData }

      // Merge existing group data
      const existingGroupData = existingInstitutionData.groups[groupId] || {}
      const currentGroupData = currentInstitutionData.groups[groupId] || {
        id: groupId,
        balance: 0,
        accountIds: []
      }
      const newGroupData = { ...existingGroupData, ...currentGroupData }

      return {
        ...result,
        [institution]: {
          ...newInstitutionData,
          balance: currentInstitutionData.balance + currentBalance.localCurrency,
          groups: {
            ...newInstitutionData.groups,
            [groupId]: {
              type: 'default',
              ...newGroupData,
              balance: currentGroupData.balance + currentBalance.localCurrency,
              accountIds: [...newGroupData.accountIds, id]
            }
          }
        }
      }
    }, {})

  // But now the institutions are not sorted and we lost existing institution data
  return Object.keys(groupedAccountIds)
    // Sort institutions by name
    .sort((a, b) => (a > b))
    // Merge new sorted institutions with existing institution data
    .reduce((result, institution) => ({
      ...result,
      [institution]: {
        ...byInstitution[institution],
        ...groupedAccountIds[institution]
      }
    }), {})
}

let accounts
let institutionData
export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_ACCOUNTS:
      return payload || initialState
    case types.CREATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.UPDATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.DELETE_ACCOUNT:
      const { [payload]: _, ...rest } = state.byId
      return {
        ...state,
        byId: rest
      }
    case types.GROUP_BY_INSTITUTION:
      return {
        ...state,
        byInstitution: groupByInstitution(state)
      }
    case types.CREATE_ACCOUNT_GROUP:
    case types.UPDATE_ACCOUNT_GROUP:
      if (payload === undefined) return state
      const { institution, accountGroup } = payload
      institutionData = state.byInstitution[institution] || { groups: {} }
      return {
        ...state,
        byInstitution: {
          ...state.byInstitution,
          [institution]: {
            ...institutionData,
            groups: {
              ...institutionData.groups,
              [accountGroup.id]: accountGroup
            }
          }
        }
      }
    default:
      return state
  }
}
