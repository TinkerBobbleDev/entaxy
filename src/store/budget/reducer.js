/* eslint-disable no-case-declarations */
import uuid from 'uuid/v4'
import types from './types'
import budgetCategories from '../../data/budgetCategories'

// https://projects.susielu.com
// http://repec.sowi.unibe.ch/stata/palettes/index.html
export const defaultColours = [
  '#1f78b4', '#b2df8a', '#e31a1c',
  '#ff7f00', '#cab2d6', '#a6cee3',
  '#33a02c', '#6a3d9a', '#fb9a99',
  '#fdbf6f', '#b15928', '#e91e63',
  '#2196f3', '#009688', '#3e2723'
]


export const generateCategoryTree = (categoriesById) => {
  const parents = Object.values(categoriesById).reduce((result, parentCategory) => {
    if (!('parentId' in parentCategory)) {
      return {
        ...result,
        [parentCategory.id]: {
          id: parentCategory.id,
          label: parentCategory.name,
          isIncome: parentCategory.isIncome,
          options: []
        }
      }
    }
    return result
  }, {})
  Object.values(categoriesById).forEach((category) => {
    if (category.parentId in parents) {
      parents[category.parentId].options.push({
        id: category.id,
        label: category.name,
        value: category.name,
        colour: category.colour
      })
    }
  })
  return Object.values(parents)
    .sort((a, b) => ((a.label.toLowerCase() > b.label.toLowerCase()) ? 1 : -1))
}

const generateInitialState = () => {
  let count = 0
  const categoriesById = Object.keys(budgetCategories).reduce((result, groupName) => {
    const groupId = uuid()
    const group = {
      id: groupId,
      name: groupName,
      isIncome: groupName === 'Income'
    }
    const categories = budgetCategories[groupName].reduce((res, categoryName) => {
      const category = {
        id: uuid(),
        parentId: groupId,
        name: categoryName,
        colour: defaultColours[count % defaultColours.length]
      }
      count += 1
      return { ...res, [category.id]: category }
    }, {})

    return {
      ...result,
      [group.id]: group,
      ...categories
    }
  }, {})

  return {
    categoriesById,
    categoryTree: generateCategoryTree(categoriesById),
    rules: {} // rules have the format => match: { category: 'cat 1', type: 'exact_match', count: 0 }
  }
}
export const initialState = generateInitialState()
let categoriesById

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_BUDGET:
      return payload || initialState
    case types.CREATE_CATEGORY:
      let colour = null
      if ('parentId' in payload) {
        // Select the next colour of the category
        const group = state.categoryTree.find((cat => cat.id === payload.parentId))
        const lastColour = group.options.length === 0
          ? defaultColours[defaultColours.length - 1]
          : group.options[group.options.length - 1].colour
        const lastColourIndex = defaultColours.indexOf(lastColour)
        colour = defaultColours[(lastColourIndex + 1) % defaultColours.length]
      }
      categoriesById = {
        ...state.categoriesById,
        [payload.id]: {
          ...payload,
          ...(colour !== null ? { colour } : {})
        }
      }
      return {
        ...state,
        categoriesById,
        categoryTree: generateCategoryTree(categoriesById)
      }
    case types.UPDATE_CATEGORY:
      categoriesById = { ...state.categoriesById, [payload.id]: payload }
      return {
        ...state,
        categoriesById,
        categoryTree: generateCategoryTree(categoriesById)
      }
    case types.DELETE_CATEGORY:
      // Remove the selected category
      const { [payload]: category, ...rest } = state.categoriesById
      // Delete all the sub categories (if they exist)
      categoriesById = Object.values(rest).reduce((res, child) => {
        if (child.parentId === category.id) {
          return res
        }
        return { ...res, [child.id]: child }
      }, {})
      // Delete all related rules
      const rules = Object.keys(state.rules).reduce((res, match) => {
        if (state.rules[match].categoryId in categoriesById) {
          return { ...res, [match]: state.rules[match] }
        }
        return res
      }, {})
      return {
        ...state,
        categoriesById,
        categoryTree: generateCategoryTree(categoriesById),
        rules
      }
    case types.CREATE_EXACT_RULE:
      return {
        ...state,
        rules: {
          ...state.rules,
          [payload.match]: {
            categoryId: payload.categoryId,
            type: 'exact_match',
            count: 0
          }
        }
      }
    case types.DELETE_EXACT_RULE:
      const r = Object.keys(state.rules).reduce((result, match) => {
        if (match !== payload) {
          return {
            ...result,
            rules: { ...result.rules, [match]: state.rules[match] }
          }
        }
        return result
      }, { ...state, rules: {} })
      return r
    case types.COUNT_RULE_USAGE:
      // reset counters
      const newState = Object.keys(state.rules).reduce((result, match) => (
        {
          ...result,
          rules: {
            ...result.rules,
            [match]: { ...result.rules[match], count: 0 }
          }
        }
      ), state)

      // count the transactions
      for (const transaction of payload) {
        if (transaction.description in newState.rules) {
          newState.rules[transaction.description].count += 1
        }
      }
      // remove unused rules
      return Object.keys(newState.rules).reduce((result, match) => {
        if (newState.rules[match].count > 0) {
          return {
            ...result,
            rules: { ...result.rules, [match]: newState.rules[match] }
          }
        }
        return result
      }, { ...newState, rules: {} })
    default:
      return state
  }
}
