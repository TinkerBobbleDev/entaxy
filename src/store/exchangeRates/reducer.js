import { omit } from 'lodash'
import parse from 'date-fns/parse'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_EXCHANGE_RATES:
      return action.payload || initialState
    case types.UPDATE_EXCHANGE_RATES:
      // payload = {
      //   '2019-01-01': { GBP: 1.4, CAD: 1.0 },
      //   '2019-02-02': { GBP: 1.3}
      // }
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(action.payload)
        .reduce((res1, date) => {
          return Object.keys(action.payload[date])
            .reduce((res2, currency) => ({
              ...res2,
              [currency]: {
                ...res2[currency],
                [parse(date).getTime()]: action.payload[date][currency]
              }
            }), res1)
        }, state)
      Object.keys(newState).forEach((currency) => {
        newState[currency].dates = Object.keys(newState[currency])
          .filter(c => c !== 'dates')
          .sort((a, b) => b - a)
      })
      return newState
    case types.DELETE_CURRENCIES:
      return omit(state, action.payload)
    default:
      return state
  }
}
