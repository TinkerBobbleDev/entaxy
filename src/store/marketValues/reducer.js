import _ from 'lodash'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_MARKET_VALUES:
      return action.payload || initialState
    case types.CREATE_MARKET_VALUE:
      return {
        ...state,
        [action.payload.ticker]: {
          marketValue: action.payload.marketValue,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.UPDATE_MARKET_VALUE:
      return {
        ...state,
        [action.payload.ticker]: {
          marketValue: action.payload.marketValue,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.DELETE_MARKET_VALUES:
      return _.omit(state, action.payload)
    default:
      return state
  }
}
