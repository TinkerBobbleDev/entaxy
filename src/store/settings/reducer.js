import _ from 'lodash'
import types from './types'

export const initialState = {
  portfolioFilters: {
    source: {},
    account: {}
  }
}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_SETTINGS:
      return action.payload || initialState
    case types.UPDATE_PORTFOLIO_FILTER_VALUE:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            [action.payload.option]: action.payload.value
          }
        }
      }
    case types.CREATE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            ...action.payload.options
          }
        }
      }
    case types.DELETE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: _.omit(state.portfolioFilters[action.payload.filterName], action.payload.options)
        }
      }
    default:
      return state
  }
}
