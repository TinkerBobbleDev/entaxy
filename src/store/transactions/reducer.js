import _ from 'lodash'
import types from './types'

export const initialState = []

export default (state = initialState, action) => {
  let index = null

  switch (action.type) {
    case types.LOAD_TRANSACTIONS:
      return action.payload || initialState
    case types.CREATE_TRANSACTION:
      return [...state, action.payload]
    case types.DELETE_TRANSACTION:
      index = _.findIndex(state, transaction => transaction.id === action.payload)
      return [...state.slice(0, index), ...state.slice(index + 1)]
    default:
      return state
  }
}
