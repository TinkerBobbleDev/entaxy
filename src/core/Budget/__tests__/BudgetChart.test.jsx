import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import BudgetChart from '../BudgetChart'
import { initialState as transactionsInitialState } from '../../../store/transactions/reducer'
import { initialState as settingsInitialState } from '../../../store/settings/reducer'
import { initialState as budgetInitialState } from '../../../store/budget/reducer'

describe('BudgetChart', () => {
  it('matches snapshot with no used categories', () => {
    const mockStore = configureMockStore()
    const store = mockStore({
      transactions: transactionsInitialState,
      settings: settingsInitialState,
      budget: budgetInitialState
    })
    const wrapper = mount((
      <Provider store={store}>
        <BudgetChart />
      </Provider>
    ))
    expect(wrapper.debug()).toMatchSnapshot()
  })
})
