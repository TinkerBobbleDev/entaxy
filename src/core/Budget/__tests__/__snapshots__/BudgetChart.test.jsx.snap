// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetChart matches snapshot with no used categories 1`] = `
"<Provider store={{...}}>
  <BudgetChart>
    <ResponsiveContainer width=\\"100%\\" height=\\"100%\\" debounce={0}>
      <div id={[undefined]} className=\\"recharts-responsive-container\\" style={{...}}>
        <LineChart data={{...}} margin={{...}} layout=\\"horizontal\\" stackOffset=\\"none\\" barCategoryGap=\\"10%\\" barGap={4} reverseStackOrder={false} width={0} height={0} />
        <ResizeDetector handleWidth={true} handleHeight={true} onResize={[Function]} skipOnMount={false} refreshRate={1000} refreshMode={[undefined]} resizableElementId=\\"\\">
          <div style={{...}} />
        </ResizeDetector>
      </div>
    </ResponsiveContainer>
  </BudgetChart>
</Provider>"
`;
